### DISCUSSION TITLE
[feature] Why does releasing packages exclude macro `NLOHMANN_DEFINE_TYPE_INTRUSIVE_WITH_DEFAULT`?

### DISCUSSION THREAD

**Discussion by fallending** [nlohmann/json#3429] [Q&A] [answered]
1. The thing is I'm using njson with CPM:
```
CPMAddPackage("gh:nlohmann/json#v3.10.5")
```

2. And generally my struct has more fields than json str, but type matched...Here I very much need :
```
NLOHMANN_DEFINE_TYPE_INTRUSIVE_WITH_DEFAULT(type, v1, v2, ...)
```

any advices? 🙏

**Comment by falbrechtskirchinger** [comment]
I'm unfamiliar with CPM. Please provide a small self-contained example.

**Reply by fallending** [reply]
Em, forget about CPM.cmake. After download source from https://github.com/nlohmann/json/archive/refs/tags/v3.10.5.zip, cant find macro `NLOHMANN_DEFINE_TYPE_INTRUSIVE_WITH_DEFAULT `~

why?

**Reply by falbrechtskirchinger** [reply]
Did you try ~3.10.6~ (that hasn't been released yet, duh)? I think that macro is somewhat new? I'll have a look.

Edit: You'll have to use the develop branch. The macro was added here: eec79d4e8a306d0dc389d6ab7294478249069e51

**Reply by fallending** [reply]
ok, i misunderstood by:
```
/*!
@brief macro
@def NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE
@since version 3.9.0
*/
#define NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Type, ...)  \
    inline void to_json(nlohmann::json& nlohmann_json_j, const Type& nlohmann_json_t) { NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_TO, __VA_ARGS__)) } \
    inline void from_json(const nlohmann::json& nlohmann_json_j, Type& nlohmann_json_t) { NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_FROM, __VA_ARGS__)) }

#define NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(Type, ...)  \
    inline void to_json(nlohmann::json& nlohmann_json_j, const Type& nlohmann_json_t) { NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_TO, __VA_ARGS__)) } \
    inline void from_json(const nlohmann::json& nlohmann_json_j, Type& nlohmann_json_t) { Type nlohmann_json_default_obj; NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_FROM_WITH_DEFAULT, __VA_ARGS__)) }
```

code above only set imported version as `3.9.0`...Ha, my bad...
Thx for ur reply..🙏
感谢

**Comment by falbrechtskirchinger** [comment, marked as answer]
The macro is not defined in version 3.10.5 but will be part of the 3.10.6 release.

**Reply by nlohmann** [reply]
I guess I should add notes to the documentation since when the macros have been introduced.

**Reply by nlohmann** [reply]
See https://github.com/nlohmann/json/pull/3431
